### PR TITLE
feat: handle incorrect presentationTimeOffset in live streams

### PR DESCRIPTION
**Pull request by caridley**
Close https://github.com/shaka-project/shaka-player/issues/3632

## Problem description
Live streams sometimes have incorrect presentationTimeOffset because packager and encoder are not in sync.
Other players can still play these streams, but shaka cannot.

## Code Change
When fetching init segments parse the timescale which is needed to compute actual media segment presentation times.

When we see a change in the value of segmentReference.timestampOffset parse moof.traf.tfdt.baseMediaDecodeTime and then compare the actual segment start time ((baseMediaDecodeTime / timescale) + segmentReference.timestampOffset )with segmentReference.startTime. 

If that difference is greater than config value streaming.maxTimestampDiscrepancy then adjust reference.timestampOffset to compensate. Save this adjustment for use on other media segments until a change in the original segmentReference.timestampOffset triggers fresh comparison of expected and actual segment start times.

This feature is controlled with these settings

- streaming.correctTimestampOffset - when true perform this correction
- streaming.maxTimestampDiscrepancy - when the discrepancy is less than this value, use the original value

### CODE REVIEW

File: externs/shaka/player.js [hunk 1]
```diff
@@ -986,6 +988,10 @@ shaka.extern.ManifestConfiguration;
  *   The maximum time a variant can be disabled when NETWORK HTTP_ERROR
  *   is reached, in seconds.
  *   If all variants are disabled this way, NETWORK HTTP_ERROR will be thrown.
+ * @property {boolean} correctTimestampOffset
+ *   Correct timestamp offsets when they appear to be incorrect
```
Comment (avelad) [inline externs/shaka/player.js:992]: You have to document what the default value is here.

File: externs/shaka/player.js [hunk 2]
```diff
@@ -986,6 +988,10 @@ shaka.extern.ManifestConfiguration;
...
+ * @property {number} maxTimestampDiscrepancy
+ *   Correct timestamp offset when greater than maxTimestampDiscrepancy
```
Comment (avelad) [inline externs/shaka/player.js:994]: You have to document what the default value is here.

File: lib/player.js [hunk 3]
```diff
@@ -6550,7 +6567,6 @@ shaka.Player = class extends shaka.util.FakeEventTarget {
   }
 };
 
-
```
Comment (avelad) [inline lib/player.js:6553]: Please revert this change.

**Review by avelad — CHANGES_REQUESTED** [review]
I have seen that you are parsing MDHD for the time scale in initSegmentReference, I think this is also done in https://github.com/shaka-project/shaka-player/pull/4389, so we should think about unifying this , so as not to be parsing MP4 in a duplicate way. @dave-nicholas

**Comment by caridley** [comment]
> I have seen that you are parsing MDHD for the time scale in initSegmentReference, I think this is also done in #4389, so we should think about unifying this , so as not to be parsing MP4 in a duplicate way. @dave-nicholas

PR #4387 is currently parsing the timescale from the media segment instead of the init segment, and does not recognize possibility of different timescales for different tracks. We have seen streams in which the timescale is available only in the init segments.

Note also that the mp4_cea_parser.js has code that parses timescale from the init segment in the same way that we are in the timestamp_offset_corrector. I copied that code from that location since it was no readily reused without refactoring.

 Perhaps we should create a new file util/mp4_segment_parsers.js which would have the common code in a function pase_init_segment_timescales() which would return a Map<number,number> mapping tracks to timescales.

File: externs/shaka/player.js (hunk 1 above)
Comment (caridley) [reply externs/shaka/player.js:992]: done

File: externs/shaka/player.js (hunk 2 above)
Comment (caridley) [reply externs/shaka/player.js:994]: done

File: lib/player.js (hunk 3 above)
Comment (caridley) [reply lib/player.js:6553]: done

**Review by avelad — APPROVED** [review]

**Comment by joeyparrish** [comment]
This is an interesting fix, and I'm not against it.

But for the sake of curiosity, what would happen if we instead had an option to output `sequenceMode: true` from the DASH parser?  It should be easy enough to hard-code `sequenceMode: true` and load the out-of-sync content to see what happens.  @caridley, can you try that experiment for us?

Of course, an option for that in DASH would not be usable on platforms that don't support sequence mode (Tizen 2, Tizen 3, and WebOS 3 as far as we are aware).  But I'd still like to know if it's viable.

**Comment by avelad** [comment]
@joeyparrish Note/opinion: although I am delighted with how the sequence mode works, we should not further restrict the platforms, in case the sequence mode works, it should be configurable and disabled by default for DASH.

**Comment by joeyparrish** [comment]
> @joeyparrish Note/opinion: although I am delighted with how the sequence mode works, we should not further restrict the platforms, in case the sequence mode works, it should be configurable and disabled by default for DASH.

Completely agreed.

**Comment by avelad** [comment]
@caridley can you rebase to resolve the errors in the tests? Thanks!

**Review by joeyparrish — CHANGES_REQUESTED** [review]
For the sake of curiosity, what would happen if we instead had an option to output `sequenceMode: true` from the DASH parser? It should be easy enough to hard-code `sequenceMode: true` and load the out-of-sync content to see what happens. @caridley, can you try that experiment for us?

Also, could you please merge from main to fix the test failure?

Thanks!

**Comment by avelad** [comment]
I am working on a PR to support sequence mode in DASH

**Comment by joeyparrish** [comment]
I've been reviewing, but I haven't finished this yet.  I hope to have a review done by EOD, and to merge if there are no issues.

File: test/media/timestamp_offset_corrector_unit.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+  describe('checkTimestampOffset()', () => {
+    it('does not alter correct timestampOffset', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected =
+          tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(corrected === false);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:81]: This isn't doing anything.  You need to write something like expect(corrected).toBe(false).

What you've written is equivalent to expect(false) or expect(true), but that's an incomplete statement from Jasmine's point of view.  It says "here's a value to test", but not what the expectation is.

File: test/media/timestamp_offset_corrector_unit.js [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+  beforeEach(() => {
+    onEvent = jasmine.createSpy('onEvent');
+    tsoc = new shaka.media.TimestampOffsetCorrector(
+        shaka.test.Util.spyFunc(onEvent));
+  });
+
+  describe('checkTimestampOffset()', () => {
+    it('does not alter correct timestampOffset', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:75]: Please add some blank lines to break these test bodies up and make them easier to read.  One convention is something like:  setup, blank line, execution, blank line, expectations.  I tend to follow something in the spirit of this, though some longer tests have more than one execute/expect phase.

File: test/media/timestamp_offset_corrector_unit.js [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+    it('corrects timestampOffset that is off by more than ' +
+        'maxTimestampDiscrepancy', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 30);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected =
+          tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(corrected);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:95]: .toBe(true)

File: test/media/timestamp_offset_corrector_unit.js [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+    it('does not correct timestampOffset that is off by less than ' +
+          'maxTimestampDiscrepancy', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 5);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected =
+          tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(corrected === false);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:114]: .toBe(false)

File: test/media/timestamp_offset_corrector_unit.js [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+    it('returns false when timestampOffset has already been corrected', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 15);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(onEvent).toHaveBeenCalled();
+      const corrected =
+          tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(corrected === false);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:129]: .toBe(false)

File: test/media/timestamp_offset_corrector_unit.js [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+    });
+
+    it('does not correct timestampOffset when disabled', () => {
+      tsoc.configure(createStreamingConfig(false, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 30);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected =
+          tsoc.checkTimestampOffset(contentType, segRef, mediaSegment);
+      expect(corrected === false);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:140]: .toBe(false)

File: test/media/timestamp_offset_corrector_unit.js [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+  describe('correctTimestampOffset()', () => {
+    it('updates timestampOffset in reference with same original ' +
+      'timestampOffset', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 30);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected = tsoc.checkTimestampOffset(
+          contentType, segRef, mediaSegment);
+      expect(corrected);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:156]: .toBe(true)

File: test/media/timestamp_offset_corrector_unit.js [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+
+    it('does not update timestampOffset in reference with different ' +
+      'original timestampOffset', () => {
+      tsoc.configure(createStreamingConfig(true, 10));
+      const segRef =
+          createSegmenRefWithTimestampOffset(-baseMediaDecodeTimeSec + 30);
+      tsoc.parseTimescalesFromInitSegment(contentType, initSegment);
+      const corrected = tsoc.checkTimestampOffset(
+          contentType, segRef, mediaSegment);
+      expect(corrected);
```
Comment (joeyparrish) [inline test/media/timestamp_offset_corrector_unit.js:172]: .toBe(true)

**Review by joeyparrish — CHANGES_REQUESTED** [review]

File: lib/media/timestamp_offset_corrector.js [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+
+    // Fields that are found in MOOF boxes
+    let baseMediaDecodeTime = 0;
+    let trackId = 0;
+
+    new Mp4Parser()
+        .box('moof', Mp4Parser.children)
+        .box('traf', Mp4Parser.children)
+        .fullBox('tfhd', (box) => {
+          if (box.flags !== null) {
```
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/timestamp_offset_corrector.js:186]: I get these build errors If we don't have those checks

/Users/P2134253/charter/shaka-player/lib/media/timestamp_offset_corrector.js:187:30: ERROR - [JSC_TYPE_MISMATCH] actual parameter 2 of shaka.util.Mp4BoxParsers.parseTFHD does not match formal parameter
found   : (null|number)
required: number
  187|               box.reader, box.flags);
                                     ^^^^^

/Users/P2134253/charter/shaka-player/lib/media/timestamp_offset_corrector.js:192:30: ERROR - [JSC_TYPE_MISMATCH] actual parameter 2 of shaka.util.Mp4BoxParsers.parseTFDT does not match formal parameter
found   : (null|number)
required: number
  192|               box.reader, box.version);
                                     ^^^^^^^

File: lib/media/timestamp_offset_corrector.js [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+   * Parses the baseMediaDecodeTime from a media segment and
+   * converts it to seconds using timescale from the
+   * associated init segment.
+   *
+   * @param {!shaka.media.TimestampOffsetCorrector.MediaData_} mediaData
+   * @param {BufferSource} segmentData
+   * @return {number} base media decode time in seconds
+   * @private
+   */
+  parseBaseMediaDecodeTime_(mediaData, segmentData) {
```
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/timestamp_offset_corrector.js:175]: will do

File: test/media/timestamp_offset_corrector_unit.js (hunk 5 above)
Comment (caridley) [reply test/media/timestamp_offset_corrector_unit.js:88]: done

File: lib/media/timestamp_offset_corrector.js [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+   * @param {shaka.util.ManifestParserUtils.ContentType} contentType
+   *   The contentType of the segment reference e.g. 'audio', 'video'
+   * @param {shaka.media.SegmentReference} reference
+   *   A media segment reference
+   * @param {BufferSource} segmentData
+   *   The data for a media segment
+   * @return {boolean} true if the timestampOffset has not already been
+   *   corrected and we detect a discrepancy in the timestampOffset.
+   */
+  checkTimestampOffset(contentType, reference, segmentData) {
```
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/timestamp_offset_corrector.js:111]: I have added some comments explaining the need for the two functions. I am not very happy with the functions names either especially with `correctTimestampOffset` also being the name of a config field.

What do you think about changing the function name `correctTimestampOffset` to `applyTimestampOffsetCorrection`?  Please let me know if you prefer something else.

**Comment by avelad** [comment]
@caridley there are some lint errors, can you fix it? Thanks!

File: lib/media/streaming_engine.js [hunk 15]
```diff
@@ -1231,6 +1240,9 @@ shaka.media.StreamingEngine = class {
 
     mediaState.performingUpdate = true;
 
+    this.timestampOffsetCorrector_.correctTimestampOffset(
```
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/streaming_engine.js]: I have added code to prevent use of this correction with sequence mode

File: lib/media/timestamp_offset_corrector.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/timestamp_offset_corrector.js]: In the end changed name of checkTimestampOffset to checkAndCorrectTimestampOffset to better reflect what that function does.

File: lib/media/timestamp_offset_corrector.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (caridley) [reply lib/media/timestamp_offset_corrector.js]: done

File: test/media/timestamp_offset_corrector_unit.js [hunk 16]
```diff
@@ -0,0 +1,181 @@
...
+
+  // Actual base media decode time for the media segment
+  const baseMediaDecodeTimeSec = 16628739.423377778;
+
+  const initSegmentReference = new shaka.media.InitSegmentReference(
+      () => [initSegmentUri], // uri
+      0, // startByte
+      null); // endByte
+
+  const createSegmenRefWithTimestampOffset = (timestampOffset) => {
```
[the remark this answers is not in the document]
Comment (caridley) [reply test/media/timestamp_offset_corrector_unit.js]: done

File: test/media/timestamp_offset_corrector_unit.js (hunk 7 above)
Comment (caridley) [reply test/media/timestamp_offset_corrector_unit.js]: done

File: test/media/timestamp_offset_corrector_unit.js (hunk 8 above)
Comment (caridley) [reply test/media/timestamp_offset_corrector_unit.js]: done

File: test/media/timestamp_offset_corrector_unit.js (hunk 9 above)
Comment (caridley) [reply test/media/timestamp_offset_corrector_unit.js]: done

**Comment by joeyparrish** [comment]
I took the liberty of fixing the latest merge conflicts (caused by a PR just merged this morning).  I'll re-review today.

File: lib/media/timestamp_offset_corrector.js [hunk 17]
```diff
@@ -0,0 +1,194 @@
...
+  checkAndCorrectTimestampOffset(contentType, reference, segmentData) {
+    const mediaData = this.getMediaData_(contentType);
+
+    // If timestampOffset has already been corrected do nothing.
+    if (mediaData.timestampOffset !== null &&
+        mediaData.timestampOffset === reference.timestampOffset) {
+      return false;
+    }
+
+    const segmentStartTime =
```
Comment (joeyparrish) [inline lib/media/timestamp_offset_corrector.js:120]: If I'm not mistaken, this comes out to 0 for all non-MP4 content, which would trigger a large correction for each segment.

Can you please skip this new functionality for other content types?  I would hate to see this break all WebM content, for example.

**Review by joeyparrish — CHANGES_REQUESTED** [review]

**Comment by caridley** [comment]
@joeyparrish  We recently changed direction on this, because we ran into another segment time problem in some live streams that could not easily be addressed with this approach. Some live streams were showing large discrepancies in baseMediaDecodeTime for short periods of time (~ 4 segments). Once again other video players (non MSE) on other devices were not having a problems, presumably because they use "sequence mode" style of playback. 

To deal with this new problem of transient discrepancies in segment vs manifest times as well as the longer term discrepancies we have chosen to update the baseMediaDecodeTime in the media segments instead of changing the timestampOffset. If there is interest in the new approach I can submit a new PR.

FYI we are not certain that sequence mode is a panacea for this kind of problem because of it's susceptibility to AV sync issues.

This PR can be closed.
